itests/kit/node_unmanaged.go

```diff
@@ -464,332 +685,424 @@ func (tm *TestUnmanagedMiner) SubmitProveCommit(
+		recordPostOrError := func(post windowPost) {
+			head, err := tm.FullNode.ChainHead(tm.ctx)
+			if err == nil {
+				post.Epoch = head.Height()
 			}
-			select {
-			case respCh <- send:
-			case <-ctx.Done():
-			default:
+			if post.Error != nil && tm.ctx.Err() == nil {
+				tm.log("WindowPoSt submission failed for sectors %v at epoch %d: %s", post.Posted, post.Epoch, post.Error)
+			} else if tm.ctx.Err() == nil {
+				postCount++
+				tm.log("WindowPoSt loop completed post %d at epoch %d for sectors %v", postCount, post.Epoch, post.Posted)
 			}
+			tm.postsLk.Lock()
+			tm.posts = append(tm.posts, post)
+			tm.postsLk.Unlock()
 		}
 
-		var postCount int
-		for ctx.Err() == nil {
-			currentEpoch, nextPost, err := tm.calculateNextPostEpoch(ctx, sectorNumber)
-			tm.t.Logf("Activating sector %d, next post %d, current epoch %d", sectorNumber, nextPost, currentEpoch)
+		if err != nil {
+			recordPostOrError(windowPost{Error: err})
+			return
+		}
+		for tm.ctx.Err() == nil {
+			postSectors, err := tm.sectorsToPost()
 			if err != nil {
-				writeRespF(err)
+				recordPostOrError(windowPost{Error: err})
 				return
 			}
-
-			nextPost += 5 // add some padding so we're properly into the window
-
-			if nextPost > currentEpoch {
-				if _, err := tm.FullNode.WaitTillChainOrError(ctx, HeightAtLeast(nextPost)); err != nil {
-					writeRespF(err)
+			tm.log("WindowPoST sectors to post in this challenge window: %v", postSectors)
+			if len(postSectors) > 0 {
+				// something to post now
+				if err = tm.submitWindowPost(postSectors); err != nil {
+					recordPostOrError(windowPost{Error: err})
 					return
 				}
+				recordPostOrError(windowPost{Posted: postSectors})
 			}
 
-			err = tm.submitWindowPost(ctx, sectorNumber, sealedCid, sealedPath, cacheDir)
-			writeRespF(err) // send an error, or first post, or nothing if no error and this isn't the first post
-			if err != nil {
+			// skip to next challenge window
+			if err := tm.advancePostLoop(); err != nil {
+				recordPostOrError(windowPost{Error: err})
 				return
 			}
-			postCount++
-			tm.t.Logf("Sector %d: WindowPoSt #%d submitted", sectorNumber, postCount)
 		}
 	}()
+}
+
+// advancePostLoop waits until we are within the next challenge window for this miner
+func (tm *TestUnmanagedMiner) advancePostLoop() error {
+	ctx, cancel := context.WithCancel(tm.ctx)
+	defer cancel()
+
+	di, err := tm.FullNode.StateMinerProvingDeadline(ctx, tm.ActorAddr, types.EmptyTSK)
+	if err != nil {
+		return fmt.Errorf("advancePostLoop: failed to get proving deadline: %w", err)
+	}
+	currentDeadlineIdx := CurrentDeadlineIndex(di)
+	nextDeadlineEpoch := di.PeriodStart + di.WPoStChallengeWindow*abi.ChainEpoch(currentDeadlineIdx+1)
+
+	tm.log("Window PoST waiting until next challenge window, currentDeadlineIdx: %d, nextDeadlineEpoch: %d", currentDeadlineIdx, nextDeadlineEpoch)
+
+	heads, err := tm.FullNode.ChainNotify(ctx)
+	if err != nil {
+		return err
+	}
+
+	for chg := range heads {
```

Comment:
Does `ChainNotify` close the returned channel if the context is cancelled. I think it does. But if not, we should handle that here.

Reply:
yes, it does, I went digging before I did this: https://github.com/filecoin-project/lotus/blob/f08ff209fa33fdd6840758064c294b22048d9aeb/chain/store/store.go#L295-L303